### ISSUE TITLE
Support classpath resource for custom test source in dynamic tests

### ISSUE THREAD

**Issue by sbrannen** [junit-team/junit5#1467]
## Background

In conjunction with #1178, support was added for creating a `UriSource` from a `URI` supplied for a dynamic container or dynamic test.

In addition, the new `UriSource.from(URI)` static factory method currently supports creation of a `FileSource` or `DirectorySource` if the supplied `URI` represents a file or directory in the local filesystem. Otherwise, a `DefaultUriSource` is created by the factory.

## Proposal

Introduce support for URIs that have a `classpath` scheme (e.g., `classpath:tests/script.js?line=23&column=42`). `UriSource.from(URI)` could then create a `ClasspathResourceSource`.

Note that `ClasspathResourceSource` does not implement `UriSource`; however, we could potentially refactor `ClasspathResourceSource` so that it implements `UriSource` where the implementation of `getUri()` would return the value returned by `ClasspathResourceSource.getClasspathResourceName()` prepended with `classpath:` (if not already present) and appended with the `FilePosition` converted to a proper query string (if the `FilePosition` is _present_).

## Related Issues

- #1178 

## Deliverables

- [x] Add support for creating a `ClasspathResourceSource` for a dynamic test if the custom test source `URI` uses the `classpath` scheme.

**Comment by sbrannen** [comment]
Tentatively slated for 5.3 M1 for _team discussion_

**Comment by marcphilipp** [comment]
I think `ClasspathResourceSource` should not implement `UriSource` because that would break code like the following:

```java
if (source instanceof UriSource) {
    // do sth. specific for URIs
} else if (source instanceof ClasspathResourceSource) {
    // do sth. specific for classpath resources
}
```

**Comment by sbrannen** [comment]
Yeah, I pondered that as well.

Thanks for sharing the concrete example.

**Comment by sbrannen** [comment]
_in progress_

**Comment by sbrannen** [comment]
FYI: the current work on this issue can be viewed in the following branch:

https://github.com/junit-team/junit5/commits/issues/1467-classpath-resource-as-custom-test-source

**Comment by sbrannen** [comment]
**Team Decision**: implement this feature, but `ClasspathResourceSource` should not implement `UriSource` .

**Comment by sbrannen** [comment]
This has been addressed in `master` in 5d4a5d3eb7c259e141619d8fa20fde028cf359a1.
